src/main/java/bio/terra/app/controller/SearchApiController.java

```diff
@@ -15,30 +22,63 @@
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestParam;
 
+import javax.servlet.http.HttpServletRequest;
 import javax.validation.Valid;
+import java.util.Optional;
 
 @Controller
 @Api(tags = {"search"})
 @ConditionalOnProperty(name = "features.search.api", havingValue = "enabled")
 public class SearchApiController implements SearchApi {
 
+    private final ObjectMapper objectMapper;
+    private final HttpServletRequest request;
+    private final IamService iamService;
+    private final AuthenticatedUserRequestFactory authenticatedUserRequestFactory;
+    private final SearchService searchService;
+
     @Autowired
-    public SearchApiController() {
-        // do nothing
+    public SearchApiController(
+        ObjectMapper objectMapper,
+        HttpServletRequest request,
+        IamService iamService,
+        AuthenticatedUserRequestFactory authenticatedUserRequestFactory,
+        SearchService searchService
+    ) {
+        this.objectMapper = objectMapper;
+        this.request = request;
+        this.iamService = iamService;
+        this.authenticatedUserRequestFactory = authenticatedUserRequestFactory;
+        this.searchService = searchService;
+    }
+
+    @Override
+    public Optional<ObjectMapper> getObjectMapper() {
+        return Optional.ofNullable(objectMapper);
+    }
+
+    @Override
+    public Optional<HttpServletRequest> getRequest() {
+        return Optional.ofNullable(request);
+    }
+
+    private AuthenticatedUserRequest getAuthenticatedInfo() {
+        return authenticatedUserRequestFactory.from(request);
     }
 
     @Override
     public ResponseEntity<SearchIndexModel> createSearchIndex(
         @PathVariable("id") String id,
-        @Valid @RequestBody SearchIndexRequest body
+        @Valid @RequestBody SearchIndexRequest searchIndexRequest
     ) {
-        SearchIndexModel searchIndexModel = new SearchIndexModel();
+        iamService.verifyAuthorization(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT, id, IamAction.READ_DATA);
```

Comment:
I think the correct permission is `IamAction.LINK_SNAPSHOT`
```suggestion
        iamService.verifyAuthorization(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT, id, IamAction.LINK_SNAPSHOT);
```

This is the permission `SnapshotsApiController.createSnapshot()` uses.

Reply:
Should it be reading data out of a snapshot though? We need this to be able to index that data.

Reply:
After thinking about this more, you're correct. I've switched it.